### PR TITLE
[clang][Sema] Fix diagnostic for function overloading in extern "C"

### LINKED ISSUE #80235: poor diagnostic due to overloading in `extern "C"` block
https://godbolt.org/z/Koo18frzd (GCC does a bit of a better job, treating them as redeclarations and mentioning they're C functions)

The following invalid C++ code:
```
extern "C" {
void Foo(int);
int Foo();
}
```
will give the error:
```
<source>:5:5: error: functions that differ only in their return type cannot be overloaded
    5 | int Foo();
      | ~~~ ^
<source>:4:6: note: previous declaration is here
    4 | void Foo(int);
      | ~~~~ ^
1 error generated.
```
which is profoundly unfair to the person writing the code. The error isn't anything to do with functions that differ only in their return type, and everything to do with being in an extern "C" block, where overloading does not exist.

(quoting @ssbr from an internal bug)

### PR DESCRIPTION
**Pull request by s-watanabe314**
Fixes #80235

When trying to overload a function within `extern "C"`, the diagnostic `functions that differ only in their return type cannot be overloaded` is given. This diagnostic is inappropriate because overloading is basically not allowed in the C language. However, if the redeclared function has the `((overloadable))` attribute, it should be diagnosed as `functions that differ only in their return type cannot be overloaded`.

This patch uses `isExternC()` to provide an appropriate diagnostic during the diagnostic process. `isExternC()` updates the linkage information cache internally, so calling it before merging functions can cause clang to crash. An example is declaring `static void foo()` and `void foo()` within an `extern "C"` block. Therefore, I decided to call `isExternC()` after the compilation error is confirmed and select the diagnostic message.  The diagnostic message is `conflicting types for 'func'` similar to the diagnostic in C, and `functions that differ only in their return type cannot be overloaded` if the `((overloadable))` attribute is given.

Regression tests verify that the expected diagnostics are given when trying to overload functions within `extern "C"` and when the `((overloadable))` attribute is present.

### CODE REVIEW

**Review by Sirraide — APPROVED** [review]
LGTM

Thanks for the patch! I’ll merge once CI is done with this.

**Comment by Sirraide** [comment]
Ah, actually, this is still missing a release note (in `clang/docs/ReleaseNotes.rst`, in the ‘Improvements to Clang's diagnostics’ section)

**Comment by s-watanabe314** [comment]
Thank you for comments. I added release note.

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -261,6 +261,8 @@ Improvements to Clang's diagnostics
   compilation speed with modules. This warning is disabled by default and it needs
   to be explicitly enabled or by ``-Weverything``.
 
+- Improved diagnostics when trying to overload a function in extern "C". (#GH80235)
```
Comment (Sirraide) [inline clang/docs/ReleaseNotes.rst:264]:
```suggestion
- Improved diagnostics when trying to overload a function in an ``extern "C"`` context. (#GH80235)
```

**Review by shafik — COMMENTED** [review]
This make sense to me, @dwblaikie wdyt?

**Comment by dwblaikie** [comment]
> This make sense to me, @dwblaikie wdyt?

Seems OK to me

**Review by dwblaikie — APPROVED** [review]

**Comment by s-watanabe314** [comment]
Thank you for the approval. I have been granted commit access, so I will merge it myself.

**Pull request merged**